### PR TITLE
Final GUNW 3.0 Packaging Updates

### PR DESCRIPTION
**Pull request by cmarshak**
This represents the final packaging updates requested by @dbekaert and ACCESS team. **Note**: the python files I modified were reformatted by [`ruff`](https://github.com/astral-sh/ruff) (via VS code) and so there appears to be more changes than what is done. Below, I have highlight important changes and spell-out important questions that can be answered before merging.

Here is a sample GUNW run by this branch: [link](https://gunw-development-testing.s3.us-west-2.amazonaws.com/jan-2024-final-updates/S1-GUNW-A-R-164-tops-20220212_20220131-222829-00071W_00047N-PP-3d6c-v3_0_0.nc). The command to generate the above product is:
```
OMP_NUM_THREADS=4
isce2_topsapp --secondary-scenes S1A_IW_SLC__1SDV_20220131T222803_20220131T222830_041711_04F690_8F5F \
                                 S1A_IW_SLC__1SDV_20220131T222828_20220131T222855_041711_04F690_28D7  \
              --reference-scenes S1A_IW_SLC__1SDV_20220212T222803_20220212T222830_041886_04FCA3_2B3E \
                                 S1A_IW_SLC__1SDV_20220212T222828_20220212T222855_041886_04FCA3_A3E2  \
              --frame-id 25502 \
              > topsapp_img.out 2> topsapp_img.err\
```
Here is the browse imagery:
image

**Note**: ionosphere and SET corrections are now default so the correction arguments are not needed in the above command.

## Ionosphere
@mgovorcin @dbekaert:

Please review this modification to the staging of the water mask for the ionosphere [here](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/final_packaging_updates/isce2_topsapp/localize_mask.py#L27-L40). Below are screenshots of the ionosphere layers (ionosphereBurstRamps and ionosphere, respectively) in the product (which can also be seen directly in the GUNW linked above):
image
image
I can provide additional ISCE files or you can check the folder on our shared server leffe in `/mnt/leffe-data2/cmarshak/Jan2024-topsapp-revision`.

## Global attributes for relevant mean values

@sssangha @mgovorcin @dbekaert 

In an email from David last week, we agree to have global attributes of the following:

* Mean (filtered) coherence masked with water
* Mean (filtered) coherence
* Mean perpendicular baseline*
* Mean parallel baseline*
* Mean incidence angle
* Mean azimuth angle (with ARIA convention) i.e. adding 90 degrees to ISCE2 raster

They are now as global attributes (copied from the `panoply` netcdf viewer):

```
:mean_filtered_coherence_without_water_mask = 0.3126283f; // float
:mean_filtered_coherence_with_water_mask = 0.32342613f; // float
:mean_incidence_angle = 28.281992f; // float
:mean_azimuth_angle = -99.18769836425781; // double
:mean_parallel_baseline = 79.44112223102859; // double
:mean_perpendicular_baseline = -130.66025246650725; // double
```
All the values except for the baselines can be obtained by averaging a band from the ISCE2 raster. The relevant paths (in the merged folder) and the band are captured [here](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/final_packaging_updates/isce2_topsapp/packaging.py#L42-L47). For taking the mean of geocoded ISCE2 rasters, the code is [here](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/final_packaging_updates/isce2_topsapp/packaging.py#L310-L351). I created a simple integration [test](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/final_packaging_updates/tests/test_packaging.py#L56-L69) by staging the `merged` directory in s3 and using those rasters to generate the means provided. If there are corrections or changes or concerns, please make specific use of this staged data so we can all be on the same page.

For extracting baseline data, you can see this code [here](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/final_packaging_updates/isce2_topsapp/packaging.py#L50-L76) that extracts the baselines from the relevant `topsProc.xml` file and computes an average for perpendicular and parallel baseline values (there are 6 total for each value). An integration test of the extraction process is [here](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/final_packaging_updates/tests/test_packaging.py#L31-L53) in which the baselines are extracted from this sample [`topsProc.xml`](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/final_packaging_updates/tests/test_data/topsProc.xml).

## Water Mask

The water mask used in all the workflows is the [ESA Land Cover 2021](https://aws.amazon.com/marketplace/pp/prodview-7oorylcamixxc) map (via [tile-mate](https://github.com/OPERA-Cal-Val/tile-mate)).

For the ionosphere, as noted above, the code is [here](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/final_packaging_updates/isce2_topsapp/localize_mask.py#L27-L40). For the browse imagery, it is captured [here](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/final_packaging_updates/isce2_topsapp/water_mask.py#L1-L28). For the mean value of the masked coherence, it reuses the browse masking and the relevant lines of code are found [here](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/final_packaging_updates/isce2_topsapp/packaging.py#L321-L330).

To see some improvement of the browse imagery from this PR compare it to the browse image generated the current `dev` branch: 
image

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -44,3 +44,5 @@ dependencies:
  - tqdm
  - dem_stitcher>=2.4.0
  - aiohttp  # only needed for manifest and swath download
+ - pip:
+   - tile-mate
```
Comment (jhkennedy) [inline environment.yml:48]: this is on conda-forge now, isn't it?

Comment (cmarshak) [reply environment.yml:48]: Yes - but it was slightly delayed and I was impatient - I will change before I merge.

Comment (cmarshak) [reply environment.yml]: Had to expand admissible python versions of tile-mate to `3.9` (was previously `>=3.10`). Getting better at the ASF release process! Thank you, Joe!

**Comment by cmarshak** [comment]
@mgovorcin - the water mask is obtained through ESA world cover 2021.

I wrote a package (similar to dem-stitcher) that merges the tiles. The permanent water class is what I am using.

Here is a basic example in the demo: https://github.com/OPERA-Cal-Val/tile-mate/blob/dev/notebooks/Basic_Demo.ipynb

And here is how it is used in the code: https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/final_packaging_updates/isce2_topsapp/localize_mask.py#L27-L40

**Comment by dbekaert** [comment]
@mgovorcin Please review impact in the ionospheric area. 

@cmarshak we should document the limitations of our ancillaries somewhere. i.e. DEM, water mask (max latitude bounds etc). Did we also confirm there are no gaps on the water masks etc. We had some issues with the DEM and patched it earlier too,.

**Comment by dbekaert** [comment]
I all this looks fine to me please do some testing to confirm this is all as expected.
Look at GUNW created from ASC,DESC and also for the north and southern hemisphere.

**Review by dbekaert — APPROVED** [review]
LGTM

**Comment by dbekaert** [comment]
@mgovorcin Please open these global variables and ensure this is meeting your expectations.

**Comment by cmarshak** [comment]
@dbekaert asked for the following:

1. to check whether the GUNW cubes (for angles) and comparable rasters (coherence) had similar means to those computed from the ISCE2 geocoded rasters as done for the global variable. This has been checked using the GUNW above and is now included as an integration test, comparing the respective means from the GUNW with those from the geocoded rasters output from ICSE2 and ensuring they are within acceptable range for the particular example. For coherence, the mean values are within `.01`. For angles, the means are within 1 degree.
2. Include `mean_unfiltered_coherence_with/without_water_mask` as global variables. These too are apart of the integration test from 1.

For larger scale studies, let's use `hyp3` after this has been merged.

**Comment by cmarshak** [comment]
Also, the `water_mask` attribute now looks like:

```
:water_mask = "water_mask_derived_from_esa_world_cover_2021_10m.geo";
```

**Pull request merged**